src/main/clojure/clara/rules/durability/fressian.clj

```diff
@@ -630,27 +628,27 @@
             rulebase (if maybe-base-rulebase
                        maybe-base-rulebase
                        (let [without-opts-rulebase
-                             (pform/with-thread-local-binding [d/node-id->node-cache (volatile! {})
-                                                               d/clj-struct-holder record-holder]
-                               (pform/with-thread-local-binding [d/node-fn-cache (-> (fres/read-object rdr)
-                                                                                     (reconstruct-expressions)
-                                                                                     (cond->
-                                                                                      (not read-only?) (com/compile-exprs opts)))]
-                                 (binding [*read-only* read-only?]
-                                   (assoc (fres/read-object rdr)
-                                          :node-expr-fn-lookup
-                                          (.get d/node-fn-cache)))))]
+                             (binding [d/*node-id->node-cache* (volatile! {})
```

Comment:
The use of volatile here strikes me as weird.

I don't think anything here is multithreaded (the use of ThreadLocal or binding in both cases of the code tends to imply single threading), so it seems like volatile! is used only to have a mutable reference. But ThreadLocal and bindings are both already mutable.

I dunno that it is worth messing with them on this pr, but just weird.

Reply:
I thought the same, the volatile here predates the conversion to thread local, however it appears to have had similar motives, my understanding of volatiles is they're an alternative to atoms but without the guarantees atom gives us. I can do some testing, I suspect the performance difference wouldn't be a large price to pay for the safety, but I want to test it a bit more.

Reply:
Also another thought is that I sort of want to get rid of all these binding nonsense and have the serializers manage that internally if possible, but will need to do more testing for that.